### PR TITLE
Enable EqualityInference to determine more equalities

### PR DESCRIPTION
**Pull request by erichwang**
EqualityInference will now attempt a single pass at rewriting
subexpressions in terms of other known equalities. This increases
the potential for finding additional expression relationships.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java [hunk 1]
```diff
@@ -328,8 +329,8 @@ public EqualityPartition(Iterable<Expression> scopeEqualities, Iterable<Expressi
 
     public static class Builder
     {
-        private final Map<Expression, Expression> map = new HashMap<>();
-        private final Multimap<Expression, Expression> reverseMap = HashMultimap.create();
+        private final DisjointSet<Expression> unionFindSet = new DisjointSet<>();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java]: `unionFindSet` is not very descriptive. Maybe call it "equivalences", "equalities" or something like that?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java [hunk 2]
```diff
@@ -355,55 +356,44 @@ public Builder addEquality(Expression expression)
...
             checkArgument(DeterminismEvaluator.isDeterministic(expression2), "Expression must be deterministic: " + expression2);
 
-            Expression canonical1 = canonicalize(expression1);
-            Expression canonical2 = canonicalize(expression2);
-
-            if (!canonical1.equals(canonical2)) {
-                map.put(canonical1, canonical2);
-                reverseMap.put(canonical2, canonical1);
-            }
+            unionFindSet.findAndUnion(expression1, expression2);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java]: Why is this called "findAndUnion"? Shouldn't it be just "find" or "merge"?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java [hunk 3]
```diff
@@ -355,55 +356,44 @@ public Builder addEquality(Expression expression)
...
             }
-            return expression;
-        }
 
-        private void collectEqualities(Expression expression, ImmutableSet.Builder<Expression> builder)
-        {
-            builder.add(expression);
-            for (Expression childExpression : reverseMap.get(expression)) {
-                collectEqualities(childExpression, builder);
+            for (Expression expression : map.keySet()) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java]: This is a bit hard to follow. It may be a good idea to add some comments to illustrate with an example.

**Comment by martint** [comment]
Add some tests.

Otherwise, looks good.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java (hunk 2 above)
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java]: that's just a library i'm using. if the name is a problem, we should fix it in another PR.
